src/dune_rules/lib.ml

```diff
@@ -1435,14 +1435,26 @@ end = struct
             in
             (acc_res, acc_selects, acc_re_exports)
           | Direct (loc, name) ->
-            let+ lib = resolve_dep db (loc, name) ~private_deps in
-            let acc_res =
-              let open Resolve.O in
-              let+ lib = lib
-              and+ acc_res = acc_res in
-              lib :: acc_res
-            in
-            (acc_res, acc_selects, acc_re_exports)
+            let ocaml_version = db.lib_config.ocaml_version in
+            if String.equal "bigarray" (Lib_name.to_string name)
+              && not (Ocaml_version.has_bigarray_library ocaml_version)
+            then
+              begin
+                User_warning.emit ~loc
+                  [ Pp.textf
+                      "Already in the Standard Library of OCaml %s"
```

Comment:
I think that this warning might not be actionable enough: if a project builds on 4.04+ and 5, we don't want it to warn on all the versions. Dune does not have access to the minimum supported ocaml version so I don't think there's a good way to solve that in dune.

Reply:
I think the aim of this PR, is to fix dune for OCaml 5.00 without breaking the old state of the ecosystem. I can't see the interest of taking account OCaml < 4.07.
@dra27 correct me if I'm wrong.

Reply:
let me rephrase what I mean. the difference I'm highlighting is between:
1. add a warning if the presence of `bigarray` in `(libraries)` is superfluous for the project because it requires ocaml 4.07 or later (point 2 in #5494)
2. add a warning if the presence of `bigarray` in `(libraries)` is found but discarded because of the ocaml version we're compiling with (what this PR does if I understand it correctly)

We want every warning to be actionnable: there should be a way to fix the project so that it builds cleanly with all its supported configurations.

In the case 1 above, the warning has a clear action: remove `bigarray` in this `(libraries)` field. This is not going to break anything and will make the warning go away. (but this strategy is difficult to implement within dune itself and I think would fit better in a lint tool)

In the case number 2, `bigarray` cannot be removed because doing so will break ocaml 4 builds. So this means that a project that supports both ocaml 4 and 5 is bound to have warnings in the ocaml 5 builds.

I think that the question we need to find an answer for is, how does the `(libraries)` field look like in a project that uses the `Bigarray` module and supports both OCaml 4 and OCaml 5?

> I can't see the interest of taking account OCaml < 4.07.

Not sure what you mean by this, but dune continues to support building projects with OCaml 4.02+.

Reply:
> ''I can't see the interest of taking account OCaml < 4.07.''
> Not sure what you mean by this, but dune continues to support building projects with OCaml 4.02+.

What I meant was that in the ecosystem it's well known OCaml < 4.07 has not `bigarray` library.

I get what your are highlighting. Another suggestion: Is it a good idea to avoid the warning (silent mode) ? 
In that case a project with OCaml 4.07+ config won't be confusing because resolving a package take account the compiler version at that moment if I'm not mistaken.

> In the case number 2, bigarray cannot be removed because doing so will break ocaml 4 builds.

Do you mean during the parsing of dune files or when we're compiling depends on OCaml version?
Did dune generate the same META files for different ocaml version like a project with OCaml 4 and 5 config ? (I'm not a expert in dune at all)

Reply:
@moyodiallo - that's not quite right, the bigarray library is present in 4.07 and earlier, too - what changed is that it was added to the Standard Library. The point of this change is to allow a user to specify `(libraries bigarray)` and have that work on OCaml 4.02 and 5.0 - i.e. it _adds_ the bigarray library on 4.02 (you must link with `bigarray.cm[x]a`) but is a no-op on 5.0, where `Bigarray` will be used from the Standard Library.

Reply:
Thanks @dra27 ! I think that the `Ocaml_version.has_bigarray_library` function can be renamed to reflect this - `stdlib_includes_bigarray` (with opposite meaning) seems more precise.

Reply:
> Thanks @dra27 ! I think that the `Ocaml_version.has_bigarray_library` function can be renamed to reflect this - `stdlib_includes_bigarray` (with opposite meaning) seems more precise.

@emillon  in the current PR ?

Reply:
Hmm there's already a `stdlib_includes_bigarray` in `Ocaml_version`, with a different meaning. So let's forget about this for now.